Modis

elliemay481 opened the pull request:
Includes:
- Scripts to download modis. Both .sh script to access the data and a python script to extract and store the relevant variables.
- Preprocessing script that interpolates and applies the mask, and computes various monthly and annual statistics.
- Notebook showing time series and global distribution of CF and TIWP.

Commit:
create download scripts for modis l3

Commit:
add notebook for modis cloud fraction time series

elliemay481 committed:
tidy up scripts and save time series to netcdf

elliemay481 committed:
Merge branch 'main' of https://github.com/SEE-GEO/ccic_climate_record_analysis into modis

Commit:
add modis global distributions to notebook

elliemay481 committed:
Merge branch 'modis' of https://github.com/SEE-GEO/ccic_climate_record_analysis into modis

Commit:
add description to modis time series notebook

Commit:
add to conda env

Commit:
added dependencies

Commit:
added dependencies

Commit:
added dependencies

elliemay481 committed:
merge of remote modis branch into local modis branch

elliemay481 committed:
update modis results notebook

elliemay481 commented:
Updates:
- Updated the method used to calculate the monthly mean for the time series
- Added script to process zonal means, and updated the notebook with zonal mean plots (processed zonal mean also stored as netcdf on geo-c)
- Added descriptions to the notebook on how I calculate various MODIS statistics.

adriaat commented on scripts/download_modis.py:
```diff
@@ -0,0 +1,165 @@
+import numpy as np
+
+year = 2017
+n_days = 365  # number of days to download (starts at Jan. 1)
```
I assume this was modified to 366 for leap years. In any case, any differences due to using 366 instead of 365 can be considered negligible.

This was meant to be part of the review, but I somehow managed to do it outside Ö

adriaat commented on scripts/process_modis.py:
```diff
@@ -0,0 +1,399 @@
+if __name__ == "__main__":
+
+    modis_data_dir = "/scratch/ccic_record/data/modis/"
+    mask = xr.open_dataset("../mask_24_for_modis.nc")
```
How was this file created?

elliemay481 replied:
Ah, this was created using the `interpolate_mask(mask_ds, original_ds)` function that you mention below that I didn't call. I only created the mask once, and must have accidentally removed the part of the script that did this.

adriaat commented on notebooks/cloud_fraction_time_series_MODIS.ipynb:
Good explanations, thanks.
A tip in case you don't know it. You can plot coastlines and other features directly with cartopy, instead geopandas:
```python
# Assumes imports and data, with ccrs = cartopy.crs
ax.pcolormesh(lons, lats, data, transform=ccrs.PlateCarree(), rasterized=True)
ax.coastlines() # Uses Natural Earth shapefiles, you can also pick the resolution
```
This becomes handy if you start to use different projections, grids, do not use 0º as the centre line, etc.

adriaat commented on scripts/download_modis.py:
```diff
@@ -0,0 +1,165 @@
+    folder_name = current_date.strftime(
+        "%j"
+    )  # day of the year as a folder name (e.g., '001')
+    source_url = f"https://ladsweb.modaps.eosdis.nasa.gov/archive/allData/61/MYD08_D3/{year}/{folder_name}"
```
I see this only downloaded MODIS data from the Aqua platform. What is the reason to omit Terra?

elliemay481 replied:
I went with Aqua because it's on the A-train, so corresponds better with CloudSat. I'm unsure if also including Terra would be better or not. Perhaps a more realistic representation of global trends to have both? I'm not sure how diurnal cycles would affect the comparison with the other datasets though. It again adds to the problem of calculating means from daily means, rather than L2 data, since we would need to average over the Terra mean and the Aqua mean to get an overall mean. I would have to think a bit deeper about whether this would make a big difference.

adriaat replied:
I added a note to the draft to mention its data from MODIS onboard Aqua

adriaat commented on scripts/download_modis.py:
```diff
@@ -0,0 +1,165 @@
+"""This script temporarily downloads L3 MODIS datafiles and stores the relevant variables in a netcdf.
+
+When running the script, it must be accompanied by a LAADS DAAC authentication token.
```
Can you give a pointer on how to get such a token?

elliemay481 replied:
I download the data [here](https://ladsweb.modaps.eosdis.nasa.gov/archive/allData/61/MOD08_D3).
And I obtain the token from logging on [here](https://urs.earthdata.nasa.gov), which has a link to the right hand side once you log in saying '_Generate token_'.

adriaat commented on scripts/download_modis.py:
```diff
@@ -0,0 +1,165 @@
+    "Cloud_Water_Path_Ice_Histo_Intervals", "i2", ("histo_bins",)
+)
+histo_bins[:] = np.array(
+    [5, 15, 35, 75, 125, 175, 225, 275, 325, 375, 425, 475, 750, 1500, 3000, 5000]
```
These don't match what I see in [the ATBD](https://atmosphere-imager.gsfc.nasa.gov/sites/default/files/ModAtmo/documents/L3_ATBD_C6_C61_2020_08_06.pdf), even if they are not used.

elliemay481 replied:
Again in the [file spec](https://atmosphere-imager.gsfc.nasa.gov/sites/default/files/ModAtmo/MOD08_D3_fs_3040.txt), at the very end of the page, under 'Cloud_Water_Path_Ice_Histo_Intervals'.
I do admit that it would have been much better/clearer to extract this directly from the file!

elliemay481 replied:
I think I see the bins that you saw in the ATBD, on page 116. I used the data set 08_D3 (level-3 daily means), and the bins listed in the ATBD are for the data set 06_L2 (level-2 data). I'm not sure why they use different bins, it seems strange. But I trust the file spec for the 08_D3 data has the correct bins for 08_D3.

adriaat commented on scripts/process_modis.py:
```diff
@@ -0,0 +1,399 @@
+    return interpolated_mask
+
+
+def apply_mask_to_dataset(original_ds, mask):
```
This function is not used in this script

elliemay481 replied:
I included it in the actual functions, since it only required 2 short lines of code. E.g. line 193 in process_modis.py.
I must have forgotten to remove the original function.

adriaat commented on scripts/process_modis.py:
```diff
@@ -0,0 +1,399 @@
+    return ds
+
+
+def interpolate_mask(mask_ds, original_ds):
```
This function is not used in this script

adriaat commented on scripts/process_modis.py:
```diff
@@ -0,0 +1,399 @@
+                IWP_counts = dataset.Cloud_Retrieval_Fraction_Ice_Pixel_Counts
+                IWP_counts = IWP_counts.where(IWP_counts >= 0)
+
+                CF_ice = dataset.Cloud_Retrieval_Fraction_Ice * 0.0001
```
I assume the 0.0001 factor is correct, as we talked about it :)
But I cannot find it in [the ATBD](https://atmosphere-imager.gsfc.nasa.gov/sites/default/files/ModAtmo/documents/L3_ATBD_C6_C61_2020_08_06.pdf) (maybe I'm looking in the wrong place)

elliemay481 replied:
Yes, maybe hard to find! It can be found in the file spec: https://atmosphere-imager.gsfc.nasa.gov/products/daily/file-spec
"_Cloud_Retrieval_Fraction_Ice:scale_factor = 0.0001d ;
Cloud_Retrieval_Fraction_Ice:add_offset = 0.0d ;_"

adriaat commented on scripts/process_modis.py:
```diff
@@ -0,0 +1,399 @@
+    return global_mean
+
+
+def calculate_monthly_means(dataset, variable, mask):
```
`mask` is not called inside the function

elliemay481 replied:
Sloppy mistake of not removing arguments that are no longer used, oops.

adriaat commented on scripts/process_modis.py:
```diff
@@ -0,0 +1,399 @@
+            IWP = group.Cloud_Water_Path_Ice_Mean
+            CF_ice = group.Cloud_Retrieval_Fraction_Ice
+            data_filtered = (
+                IWP.where(IWP >= 0).fillna(0) * CF_ice.where(CF_ice >= 0) * 0.0001
+            )
+        elif variable == "CF":
+            CF = group.Cloud_Fraction_Mean
+            data_filtered = CF.where(CF >= 0) * 0.0001
```
Help me out here please...
The 0.0001 scaling factor is only used once because IWP was already scaled? If yes, where was it scaled? Or there's something that operating with non-scaled data first and then scaling it afterwards is equivalent?
I assume this is correct anyway, otherwise we would have noticed differences of orders of magnitude, I suppose...

elliemay481 replied:
The scaling factor is only used once because IWP doesn't need to be scaled. From the file spec:
"_Cloud_Water_Path_Ice_Mean:scale_factor = 1.0d ;
 Cloud_Water_Path_Ice_Mean:add_offset = 0.0d ;_"

adriaat commented on scripts/process_modis.py:
```diff
@@ -0,0 +1,399 @@
+                CF_counts = CF_counts.where(CF_counts >= 0)
+
+                IWP = IWP.where(IWP >= 0).fillna(0)
+                TIWP = IWP * CF_ice
```
Same question here as before; IWP is not scaled yet it was downloaded and saved as an int32. Why doesn't IWP require explicit scaling?

elliemay481 replied:
I'm guessing because the unit is g/m^{2}, and they don't retrieve less than 1 g/m^{2}.
"_Cloud_Water_Path_Ice_Mean:units = "g/m^2" ;
Cloud_Water_Path_Ice_Mean:valid_range = 0s, 6000s ;
Cloud_Water_Path_Ice_Mean:_FillValue = -9999s ;
Cloud_Water_Path_Ice_Mean:scale_factor = 1.0d ;_"

It also comes as an int in the original data. (Actually an int16, now that I check again, so I could have saved some storage space here).

adriaat reviewed (commented):
Have a look at the comments. I don't think there's any action required from your side, but I'd like to assert a few questions before I click 'approve' :)
You can ignore a few comments, such as "this is not used"; it's part of writing this one-time only scripts.
The main comments perhaps are to leave written down:
- how did you get the mask in `"../mask_24_for_modis.nc"`,
- how to get this "LAADS DAAC authentication token",
- a link to the ATBD of the product somewhere (can even be in the notebook),
- and why the scaling factor is needed (I don't find this info in [the ATBD](https://atmosphere-imager.gsfc.nasa.gov/sites/default/files/ModAtmo/documents/L3_ATBD_C6_C61_2020_08_06.pdf), if this is the right one).

Commit:
removed unused functions in processing script and added link to modis atdb in notebook

elliemay481 commented:
Some small changes based on Adria's comments:

- Added a link in the notebook to the relevant MODIS webpage, to find ATBD/file spec/user guide/etc
- Removed unused functions and unused arguments in the file 'process_modis.py'

adriaat reviewed (approved):
Good work, also with the answers to my questions! 👍
